### DISCUSSION TITLE
Support for Multiple Languages in a Single Notebook Document

### DISCUSSION THREAD

**Discussion by samir-abis** [eclipse-langium/langium#1488] [Q&A] [answered]
Hello Langium Community,

I'm working on a project that involves using Langium to define a language for metadata and validation rules within a notebook environment. We've encountered a challenge that I believe could benefit from your insights and potential solutions.

**Problem Description:**

In our setup, we have a notebook document that can contain cells with different languages (e.g., metadata, validations, permissions). Each cell within the notebook is expected to have its own service instance based on the language specified in the cell. However, we've found that the `getServices` method in our custom service registry implementation (`CustomServiceRegistry`) is not capable of selecting the correct services for individual cells within the notebook. This limitation arises because `getServices` is designed to work with the URI of the entire notebook file, rather than the specific cell.

**Current Implementation:**

Our current approach involves creating a new instance of `CustomCompiler` for each cell to ensure each cell has its own service. However, this approach does not dynamically select the appropriate services based on the cell's language. Instead, it relies on the file extension of the notebook file, which is not sufficient for our use case.

**Proposed Solution:**

We've considered modifying the `getServices` method to accept an additional parameter for the cell's language. This way, we could dynamically select the appropriate services based on the cell's language. However, this approach requires changes to how `getServices` is called throughout our codebase, which might not be straightforward.

**Questions:**

1. Is there a recommended way to handle multiple languages within a single notebook document in Langium?
2. Are there any built-in mechanisms or best practices in Langium for dynamically selecting services based on the language of individual cells within a notebook?
3. Has anyone faced a similar challenge, and if so, how did you address it?

**Additional Context:**

- Our project is a monorepo containing a language project designed to be user-friendly for non-programmers working with it.
- We're using Langium for defining the AST and syntax of our language, along with language server support.
- The notebook environment is integrated into VS Code using a custom VS Code extension.

We're looking forward to any suggestions, insights, or potential solutions that could help us overcome this limitation. Thank you for your time and support.

**Comment by msujew** [comment]
Hey @samir-abis-everest,

> This limitation arises because `getServices` is designed to work with the URI of the entire notebook file, rather than the specific cell.

As far as I know, we don't manipulate the URI that comes from vscode in any way. Therefore, it should contain the full cell URI (which is effectively just `vscode-notebook-cell://<notebook-path>#<cell-handle>`). 

> We've considered modifying the getServices method to accept an additional parameter for the cell's language.

I'm a bit confused about this: Do you have multiple languages - that are registered as separate languages - in VSCode that use the same language server? If so, how? I'm not aware of such a functionality in the implementation of `vscode-languageclient`.

**Reply by samir-abis** [reply]
@msujew,

Thank you for your insightful response. It's clear that the URI for each cell in a notebook document is indeed unique and should be sufficient for our needs. However, the challenge we're facing is more about dynamically selecting the appropriate services based on the cell's language within our custom service registry implementation.

Regarding your question about multiple languages registered as separate languages in VSCode that use the same language server, it's indeed possible and quite common in scenarios where languages share a lot of syntax or are closely related. This is achieved by implementing the language server in a way that it can parse and understand the syntax of all the languages it supports.

```json
{
 "contributes": {
    "languages": [
      {
        "id": "languageA",
        "aliases": ["Language A", "langA"],
        "extensions": [".langA"],
        "configuration": "./language-configuration.json"
      },
      {
        "id": "languageB",
        "aliases": ["Language B", "langB"],
        "extensions": [".langB"],
        "configuration": "./language-configuration.json"
      }
    ]
 },
 "activationEvents": [
    "onLanguage:languageA",
    "onLanguage:languageB"
 ],
 "main": "./extension.js"
}
```

The language server uses the file extension or other metadata to determine which language the file belongs to and enables or disables certain features based on that.

In the context of our project, where we're using Langium to define a language for metadata and validation rules within a notebook environment, we're looking for a way to dynamically select services based on the language of individual cells within a notebook. This would allow us to provide language-specific features and validations for each cell, enhancing the user experience significantly.

Thank you for your time and support.

Best regards,
Samir Abis

**Reply by msujew** [reply]
> This is achieved by implementing the language server in a way that it can parse and understand the syntax of all the languages it supports.

I'm aware of that, I implemented the [multi language support](https://github.com/eclipse-langium/langium/pull/311). I guess I never noticed how you can register multiple languages on a single language client via the `LanguageClientOptions`. Alrighty.

My recommendation would be as follows:

1. Create an override of `ServiceRegistry` (you likely have one already) and create a `Map<string, string>` that stores an association of document URI to document language id in there.
2. Override the `getServices` method appropriately so that when a document with a known URI is requested, it will return the service instance associated with its known service instance.
3. Override the `DefaultLangiumDocumentFactory` and override the `createAsync` method like this:
```ts
protected async createAsync<T extends AstNode = AstNode>(uri: URI, content: string | TextDocument, cancelToken: CancellationToken): Promise<LangiumDocument<T>> {
  const languageId = this.textDocuments?.get(uri.toString)?.languageId;
  if (languageId) {
    this.serviceRegistry.storeLanguageId(uri.toString(), languageId);
  }
  return super.createAsync(uri, content, cancelToken);
}
```
Maybe you need to override some more services in the `DefaultLangiumDocumentFactory`, but that should work very similarly.

**Reply by cdietrich** [reply]
do we need the map or can we do it on the fly:
textDocuments.get(uri).languageId? (inside the service registry)
or would this be too expensive? or too late?

**Reply by msujew** [reply]
That should actually work as well, yeah. I initially added the map because I recommended a different (async) approach, see the history of that comment. But that probably isn't even necessary.

As the lookup is be O(1) there shouldn't be any performance impact. At this point the text documents are already filled, so it's also not too late.

**Reply by msujew** [reply]
@samir-abis-everest @cdietrich I've adapted this approach in https://github.com/eclipse-langium/langium/pull/1492, as this makes a lot of sense as the default behavior of Langium.

**Reply by cdietrich** [reply]
@msujew i have a follow up question/problem.
with the artifical uris we cannot reason on the uri in other services
in Notebooks you can enrich `NotebookCell`s with metadata.
so we could store additional information.

do you know if this information is somewhere funnelled to the server as the the cells are as text document?

if no:
when we implement `notebooks.synchronization.onDidOpenNotebookDocument`and similar the text documents are no longer mainted. so i assume we need to do this ourself

what is the proper way to have both the textdocuments populated and maintained and the langiumdocuments in the documentbuilder filled?

manually downcast and fill the map, then call update?

```
for (const cell of p1.cellTextDocuments) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
    ;((langiumSharedServices.workspace.TextDocuments as any)._syncedDocuments as Map<string, unknown>).set(
      cell.uri,
      TextDocument.create(cell.uri, cell.languageId, 1, cell.text)
    )
    uris.push(URI.parse(cell.uri))
  }
  langiumSharedServices.workspace.DocumentBuilder.update(uris, []).catch(x => {
    console.error(x)
  })
```
or is there a way to make this nicer / better adapted?

**Reply by msujew** [reply, marked as answer]
The `vscode-languageserver` package exports a `NotebookDocuments` class, that should be used similar to the currently used `TextDocuments` class. One would need to use that in the `LanguageServer` service implementation and also adapt the `LangiumDocumentsFactory` to handle that better.

It probably makes sense to implement support for that natively in Langium. I've created https://github.com/eclipse-langium/langium/issues/1496 to track this.

**Reply by cdietrich** [reply]
i guess the interesting point is to make it work "hybrid".
some normal and notebook documents in the same ws.

**Reply by msujew** [reply]
Luckily, Langium is rarely using the `TextDocuments`, and so just using a simple `TextDocuments.get(uri) ?? NotebookDocuments.get(uri)` lookup should work in most circumstances.

**Reply by cdietrich** [reply]
i see do you think we can reuse the text documents like

```
const nds = new NotebookDocuments(langiumSharedServices.workspace.TextDocuments)
nds.onDidChange(change => console.log(change))
nds.listen(connection)
```
or would the NotebookDocuments need its own text documents?

**Reply by msujew** [reply]
Uhm, haven't tried it yet, but I can imagine that this should work out well.
